internal/host/host.go

```diff
@@ -320,23 +339,34 @@ func (m *Manager) RefreshStatus(ctx context.Context, h *models.Host, db *gorm.DB
 	if err != nil {
 		return err
 	}
-	conditions, validationsResults, err := m.rp.preprocess(vc)
+	conditions, newValidationRes, err := m.rp.preprocess(vc)
 	if err != nil {
 		return err
 	}
-	if err = m.reportValidationStatusChanged(ctx, vc, h, validationsResults); err != nil {
+	currentValidationRes, validationsChanged, err := m.checkValidationChanged(ctx, h, newValidationRes)
+	if err != nil {
 		return err
 	}
 	err = m.sm.Run(TransitionTypeRefresh, newStateHost(h), &TransitionArgsRefreshHost{
 		ctx:               ctx,
 		db:                db,
 		eventHandler:      m.eventsHandler,
 		conditions:        conditions,
-		validationResults: validationsResults,
+		validationResults: newValidationRes,
 	})
 	if err != nil {
 		return common.NewApiError(http.StatusConflict, err)
 	}
+	if validationsChanged {
+		// reporting validations change in status needs to compare the new validations results
+		// with the current validations results, therefore, it should be done before updating the DB.
+		m.reportValidationStatusChanged(ctx, vc, h, newValidationRes, currentValidationRes)
+		// the transitions are using the validations info in order to choose the correct
+		// transition, therefore, we need to make sure we update them only after the transitions.
+		if _, err = m.updateValidationsInDB(ctx, h, newValidationRes); err != nil {
```

Comment:
If the host changed status during the state machine invocation, will the validations be updated?

Reply:
No, I removed the validations update from inside the transition (PostRefreshHost() won't update validations results in DB anymore)